### ISSUE TITLE
Incorrect SELECT query generation including "LIMIT" clause and "M:N" association with "where" clause returns unexpected results

### ISSUE THREAD

**Issue by fogine** [sequelize/sequelize#4716]
Hello,

I've come across a bug with query generation which include limiting results with including `belongsToMany` relation containing `WHERE` clause.

I made executable [script](https://gist.github.com/fogine/acb637ec1b4ab2ef5d81) to demonstrate this behavior as they are cases when generated query might return expected result if query and data in database are in specific state.

if you investigate the file the following are queries generated for particular demonstration cases:
## "Test 1": Works

``` javascript
    Student.findAndCount({
        offset: 6,
        order: [["id", "DESC"]],
        include: [
            {
                model: Teacher,
                as: 'teachers',
                where: {
                    id: {$in: [3, 2]}
                }
            }
        ]
    });
```

``` sql
SELECT "student"."id", "student"."age", "student"."name",
"student"."created_at", "student"."updated_at", "teachers"."id" AS
"teachers.id", "teachers"."age" AS "teachers.age", "teachers"."name" AS
"teachers.name", "teachers"."created_at" AS "teachers.created_at",
"teachers"."updated_at" AS "teachers.updated_at",
"teachers.student_teacher"."id" AS "teachers.student_teacher.id",
"teachers.student_teacher"."teacher_id" AS
"teachers.student_teacher.teacher_id", "teachers.student_teacher"."student_id"
AS "teachers.student_teacher.student_id"
FROM "student" AS "student"
INNER JOIN (
    "student_teacher" AS "teachers.student_teacher"
    INNER JOIN "teacher" AS "teachers"
    ON "teachers"."id" = "teachers.student_teacher"."teacher_id"
)
ON "student"."id"  = "teachers.student_teacher"."student_id"
AND "teachers"."id" IN (3, 2)
ORDER BY "student"."id" DESC OFFSET 6;
```
## "Test 2": Bug - Wrong query generated... but this case returns expected result by "coincidence"

``` javascript
    Student.findAndCount({
        limit: 5,
        offset: 6,
        order: [["id", "DESC"]],
        include: [
            {
                model: Teacher,
                as: 'teachers',
                where: {
                    id: {$in: [3, 2]}
                }
            }
        ]
    });
```

``` sql
SELECT "student".*, "teachers"."id" AS "teachers.id", "teachers"."age" AS
"teachers.age", "teachers"."name" AS "teachers.name", "teachers"."created_at"
AS "teachers.created_at", "teachers"."updated_at" AS "teachers.updated_at",
"teachers.student_teacher"."id" AS "teachers.student_teacher.id",
"teachers.student_teacher"."teacher_id" AS
"teachers.student_teacher.teacher_id", "teachers.student_teacher"."student_id"
AS "teachers.student_teacher.student_id"
FROM (
    SELECT "student"."id", "student"."age", "student"."name",
    "student"."created_at", "student"."updated_at"
    FROM "student" AS "student"
    WHERE (
        SELECT "student_teacher"."id"
        FROM "student_teacher" AS "student_teacher"
        INNER JOIN "teacher" AS "teacher"
        ON "student_teacher"."teacher_id" = "teacher"."id"
        WHERE ("student"."id" = "student_teacher"."student_id")
        LIMIT 1
    ) IS NOT NULL
    ORDER BY "student"."id" DESC LIMIT 5 OFFSET 6
) AS "student"
INNER JOIN (
    "student_teacher" AS "teachers.student_teacher"
    INNER JOIN "teacher" AS "teachers"
    ON "teachers"."id" = "teachers.student_teacher"."teacher_id"
)
ON "student"."id" = "teachers.student_teacher"."student_id"
AND "teachers"."id" IN (3, 2)
ORDER BY "student"."id" DESC;
```
## "Test 3": Bug - Wrong query - returns unexpected results

``` javascript
    Student.findAndCount({
        limit: 5,
        offset: 6,
        include: [
            {
                model: Teacher,
                as: 'teachers',
                where: {
                    id: {$in: [3, 2]}
                }
            }
        ]
    });
```

``` sql
SELECT "student".*, "teachers"."id" AS "teachers.id", "teachers"."age" AS
"teachers.age", "teachers"."name" AS "teachers.name", "teachers"."created_at"
AS "teachers.created_at", "teachers"."updated_at" AS "teachers.updated_at",
"teachers.student_teacher"."id" AS "teachers.student_teacher.id",
"teachers.student_teacher"."teacher_id" AS
"teachers.student_teacher.teacher_id", "teachers.student_teacher"."student_id"
AS "teachers.student_teacher.student_id"
FROM (
    SELECT "student"."id", "student"."age", "student"."name",
    "student"."created_at", "student"."updated_at"
    FROM "student" AS "student"
    WHERE (
        SELECT "student_teacher"."id"
        FROM "student_teacher" AS "student_teacher"
        INNER JOIN "teacher" AS "teacher"
        ON "student_teacher"."teacher_id" = "teacher"."id"
        WHERE ("student"."id" = "student_teacher"."student_id")
        LIMIT 1
    ) IS NOT NULL LIMIT 5 OFFSET 6
) AS "student"
INNER JOIN (
    "student_teacher" AS "teachers.student_teacher"
    INNER JOIN "teacher" AS "teachers"
    ON "teachers"."id" = "teachers.student_teacher"."teacher_id"
)
ON "student"."id" = "teachers.student_teacher"."student_id"
AND "teachers"."id" IN (3, 2);

```

As I look at queries it seems it would be great if generated queries for test cases `2` and `3` would be the same as in `1` with `LIMIT` included at the end of query, wouldn't it?

Thank you for your work!

**Comment by mickhansen** [comment]
Does it work `findAll` rather than `findAndCount`?
The `limit` is the culprit here since it triggers subquery logic.

**Comment by fogine** [comment]
No, It seems it behaves the same way. Is it working for you with `findAll`?

**Comment by mickhansen** [comment]
Haven't tested it yet, just wanted to make sure that the method was not a factor.
Unfortunately `limit` doesn't work in all cases with includes because of cartesian products.

**Comment by fogine** [comment]
well, do you mean it might be problem to generate appropriate sql query in this case? Could you be more specific in which cases it's not possible, please?
It really looks like the problem is caused by `LIMIT` here. In this case use of subquery is not neccesary, is it?

**Comment by mickhansen** [comment]
The limit is necessary when you include an association that result in a cartesian product, which would be the case for a belongsToMany association.

limit for includes is not supported in all cases as i said, but it would. In this case it looks like the `where` statement is missing from the nested filter subquery, which is highly odd.

It would be helpfull if you could provide a SSCCE - Preferably as a PR with an integration test.

**Comment by fogine** [comment]
I'm not sure you've noticed that I've made a SSCCE:
https://gist.github.com/fogine/acb637ec1b4ab2ef5d81
I think it should demonstrate the problem pretty well.
I can make PR with an integration test though.

**Comment by mickhansen** [comment]
@fogine Ah no i missed that. That looks pretty copy/pasteable into a test, so no need to provide a PR.

**Comment by markdboyd** [comment]
Is there are any progress here? I'm fairly sure that I'm running into a similar situation, where I'm using limit and a join that I need isn't being provided on the subquery

**Comment by mickhansen** [comment]
@markdboyd Not much, overall we're working on simplifying this by running associations in seperate queries.
However if you could provide a SSCCE for your case i'd be happy to take a look.

**Comment by markdboyd** [comment]
Basically, my problem is that I have a where clause on the subquery (which is triggered by a limit), but it it isn't working because it references a table that is joined outside the subquery. I'm not sure how much detail to provide for a SSCCE, but here goes: https://gist.github.com/markdboyd/cf19937dec07b1af2bd7

It also seems like #4737 is a related issue here.

**Comment by mickhansen** [comment]
Because of your custom where that is played inside the subquery, don't have a workaround for a planned fix for that at the moment.

**Comment by markdboyd** [comment]
Just so I understand, are you saying that the custom where condition itself is the problem? I was under the impression the problem was the table that needs to be joined for the custom where condition (in the subquery) is only joined on the outer query, not the subquery.

**Comment by mickhansen** [comment]
It gets places inside the subquery cause we wouldn't know not to.

**Comment by demege** [comment]
Hi, Im having the same issue.
Any progress on this topic.

**Comment by ybrodsky** [comment]
Same issue here. Any progress or workaround?

**Comment by janmeier** [comment]
@demege @ybrodsky More test cases are always welcome

**Comment by ybrodsky** [comment]
@janmeier my case is exactly the same as @fogine original post. 

I have noticed that adding the option `subQuery: false` seems to produce the desired mysql query. Is there any downside of using this?

**Comment by janmeier** [comment]
When adding offset and limit to a duplicating (hasMany or belongsToMany relations) you need a subquery - otherwise the offset and limit is applied to the cartesian product, and not the the main table

**Comment by ybrodsky** [comment]
Shouldnt it be applied to the cartesian product if it's an inner join?

My case:

```
Oportunity belongsToMany Client (through ClientOportunity)
Oportunity belongsTo User
Client belongsToMany Oportunity (through ClientOportunity)
```


```
Oportunity.findAll({
  status: {'$in': [1, 2, 0]}
  attributes: ['blablabla'],
  offset: 0,
  limit: 10,
  include: [
    {
      model: Client,
      as: 'client',
      attributes: ['blablabla']
    }
    {
      model: User,
      as: 'user',
      attributes: ['blablabla']
    }
  ]
})

```

Produced query:

```
SELECT `Oportunity`.*, `user`.`id` AS `user.id`, `user`.`name` AS `user.name`,
 `user`.`surname` AS `user.surname`, `client`.`id` AS `client.id`, 
`client`.`name` AS `client.name`, `client`.`surname` AS `client.surname`, 
`client.ClientOportunity`.`id` AS `client.ClientOportunity.id`, 
`client.ClientOportunity`.`client_id` AS `client.ClientOportunity.client_id`, 
`client.ClientOportunity`.`oportunity_id` AS `client.ClientOportunity.oportunity_id`, 
`client.ClientOportunity`.`createdAt` AS `client.ClientOportunity.createdAt`, 
`client.ClientOportunity`.`updatedAt` AS `client.ClientOportunity.updatedAt` 
FROM (
  SELECT `Oportunity`.`id`, `Oportunity`.`name`, `Oportunity`.`status`, 
`Oportunity`.`createdAt`, `Oportunity`.`user_id` FROM `oportunity` AS `Oportunity` 
  WHERE `Oportunity`.`status` IN (0, 1, 2) LIMIT 0, 10
) AS `Oportunity` 
LEFT OUTER JOIN `user` AS `user` ON `Oportunity`.`user_id` = `user`.`id` 
LEFT OUTER JOIN (
  `client_oportunity` AS `client.ClientOportunity` 
  INNER JOIN `client` AS `client` ON `client`.`id` = `client.ClientOportunity`.`client_id`
) ON `Oportunity`.`id` = `client.ClientOportunity`.`oportunity_id` 
ORDER BY id DESC
```

**Comment by ybrodsky** [comment]
I have tested further and noted two things:
- The resulting query doesn't have the `ORDER` clause inside the subquery. So the results in the subquery are messed up to start with.
- If you have conditions for your joined models, the conditions are not inside the subquery neither. So again the results in the subquery are messed up.

Testing with raw mysql, if you add the `order` clause inside the subquery and (if any) the join conditions inside the subquery, the results are as expected.

Example query which returns expected results when using limit and a condition for the joined model.

```
SELECT `Oportunity`.*, `user`.`id` AS `user.id`, `user`.`name` AS `user.name`, `user`.`surname` AS `user.surname`,
`client`.`id` AS `client.id`, `client`.`name` AS `client.name`, `client`.`surname` AS `client.surname`,
`client.ClientOportunity`.`id` AS `client.ClientOportunity.id`, `client.ClientOportunity`.`client_id` AS
`client.ClientOportunity.client_id`, `client.ClientOportunity`.`oportunity_id` AS `client.ClientOportunity.oportunity_id`,
`client.ClientOportunity`.`createdAt` AS `client.ClientOportunity.createdAt`, `client.ClientOportunity`.
`updatedAt` AS `client.ClientOportunity.updatedAt`

FROM (
  SELECT `Oportunity`.`id`, `Oportunity`.`name`, `Oportunity`.`status`, `Oportunity`.`createdAt`, `Oportunity`.`user_id`
  FROM `oportunity` AS `Oportunity`
  WHERE `Oportunity`.`status` IN (0, 1, 2)
  AND (
    SELECT `ClientOportunity`.`id`
    FROM `client_oportunity` AS `ClientOportunity`
    INNER JOIN `client` AS `Client` ON `ClientOportunity`.`client_id` = `Client`.`id`
      AND (`Client`.`email` LIKE '%cacalano%' OR `Client`.`name` LIKE '%cacalano%' OR `Client`.`surname` LIKE '%cacalano%')
    WHERE (`Oportunity`.`id` = `ClientOportunity`.`oportunity_id`) LIMIT 1
  ) IS NOT NULL ORDER BY id DESC LIMIT 0, 10
) AS `Oportunity`
LEFT OUTER JOIN `user` AS `user` ON `Oportunity`.`user_id` = `user`.`id`
INNER JOIN (`client_oportunity` AS `client.ClientOportunity`
  INNER JOIN `client` AS `client` ON `client`.`id` = `client.ClientOportunity`.`client_id`
) ON `Oportunity`.`id` = `client.ClientOportunity`.`oportunity_id`

AND (`client`.`email` LIKE '%cacalano%' OR `client`.`name` LIKE '%cacalano%' OR `client`.`surname` LIKE '%cacalano%')
ORDER BY Oportunity.id DESC
```

**Comment by mkaufmaner** [comment]
If you are having problems with sub query generation with required includes, try explicitly setting the main property `topLimit` of options.
ie.
`{topLimit: 10}`

**Comment by Alexsey** [comment]
My issue https://github.com/sequelize/sequelize/issues/7344 is looks like related. And adding of `subQuery: false` is looks like a fix for my case as well

**Comment by brianemoore57** [comment]
I ran into this because of the deletedAt issue. I want to use userName column in my User table, and as a soft-delete, preserve the option to re-use the userName if first record deleted. I had to change to a exclude constraint to get this behaviour. Then, sequelize.model inserted this options.limit since logic adds it unless a Pk, or a unique column. It doesnt see my exclude constraint as a unique.  I know the deletedAt is changed to a boolean flag, so I will be able to use a regular multi-column unique constraint. However, this logic here seems to just generate wrong SQL, at least for postgresql, limiting its use.
